### PR TITLE
db: Improve defaultRepos.List query

### PR DESCRIPTION
**Pull request by tsenart**
This commit improves the `defaultRepos.List` query to use a better plan — it results in rougly 480ms runtime in production currently, vs the previous 1.4s. It also fixes a bug with the old query (we weren't filtering out deleted repos from the default_repos table).

Thanks @felixge for the SQL hacking session :)

#### Before

```
Unique  (cost=1325604.75..1351922.72 rows=3509063 width=36) (actual time=1135.181..1410.477 rows=206416 loops=1)
  ->  Sort  (cost=1325604.75..1334377.41 rows=3509063 width=36) (actual time=1135.179..1157.578 rows=206416 loops=1)
        Sort Key: r.id, r.name
        Sort Method: external merge  Disk: 9984kB
        ->  Append  (cost=168704.21..831721.53 rows=3509063 width=36) (actual time=760.768..1046.177 rows=206416 loops=1)
              ->  Hash Semi Join  (cost=168704.21..639660.16 rows=3302671 width=39) (actual time=760.395..760.395 rows=0 loops=1)
                    Hash Cond: (r.id = sr.repo_id)
                    ->  Seq Scan on repo r  (cost=0.00..347065.21 rows=3302671 width=39) (actual time=0.006..0.006 rows=1 loops=1)
                          Filter: (deleted_at IS NULL)
                    ->  Hash  (cost=107855.17..107855.17 rows=3708883 width=4) (actual time=760.255..760.255 rows=0 loops=1)
                          Buckets: 131072  Batches: 64  Memory Usage: 1024kB
                          ->  Hash Join  (cost=1.38..107855.17 rows=3708883 width=4) (actual time=760.254..760.254 rows=0 loops=1)
                                Hash Cond: (sr.external_service_id = s.id)
                                ->  Seq Scan on external_service_repos sr  (cost=0.00..95690.83 rows=3708883 width=12) (actual time=0.008..443.226 rows=3643696 loops=1)
                                ->  Hash  (cost=1.17..1.17 rows=17 width=8) (actual time=0.013..0.013 rows=3 loops=1)
                                      Buckets: 1024  Batches: 1  Memory Usage: 9kB
                                      ->  Seq Scan on external_services s  (cost=0.00..1.17 rows=17 width=8) (actual time=0.008..0.010 rows=3 loops=1)
                                            Filter: (namespace_user_id IS NOT NULL)
                                            Rows Removed by Filter: 16
              ->  Gather  (cost=1000.85..139425.42 rows=206392 width=39) (actual time=0.371..269.532 rows=206416 loops=1)
                    Workers Planned: 2
                    Workers Launched: 2
                    ->  Nested Loop  (cost=0.85..117786.22 rows=85997 width=39) (actual time=0.096..279.144 rows=68805 loops=3)
                          ->  Parallel Index Only Scan using default_repos_pkey on default_repos  (cost=0.42..3522.35 rows=85997 width=4) (actual time=0.047..37.204 rows=68805 loops=3)
                                Heap Fetches: 206416
                          ->  Index Scan using repo_pkey on repo  (cost=0.43..1.33 rows=1 width=39) (actual time=0.003..0.003 rows=1 loops=206416)
                                Index Cond: (id = default_repos.repo_id)
Planning Time: 0.547 ms
Execution Time: 1425.366 ms
```

#### After (attempt 1)

```
Unique  (cost=493745.14..495867.59 rows=424491 width=39) (actual time=421.170..476.224 rows=206400 loops=1)
  ->  Sort  (cost=493745.14..494806.37 rows=424491 width=39) (actual time=421.169..443.556 rows=206400 loops=1)
        Sort Key: repo.id
        Sort Method: external merge  Disk: 9976kB
        ->  Gather  (cost=1000.43..447265.20 rows=424491 width=39) (actual time=0.369..321.361 rows=206400 loops=1)
              Workers Planned: 2
              Workers Launched: 2
              ->  Nested Loop  (cost=0.43..403816.10 rows=176871 width=39) (actual time=0.079..322.571 rows=68800 loops=3)
                    ->  Parallel Append  (cost=0.00..60190.47 rows=304167 width=4) (actual time=0.033..13.560 rows=68805 loops=3)
                          ->  Nested Loop  (cost=0.43..54996.19 rows=218170 width=4) (actual time=0.040..0.040 rows=0 loops=1)
                                ->  Seq Scan on external_services es  (cost=0.00..1.17 rows=1 width=8) (actual time=0.040..0.040 rows=0 loops=1)
                                      Filter: ((namespace_user_id IS NOT NULL) AND (deleted_at IS NULL))
                                      Rows Removed by Filter: 19
                                ->  Index Scan using external_service_repos_external_service_id on external_service_repos esr  (cost=0.43..45722.81 rows=927221 width=12) (never executed)
                                      Index Cond: (external_service_id = es.id)
                          ->  Parallel Seq Scan on default_repos  (cost=0.00..2128.07 rows=121407 width=4) (actual time=0.018..7.518 rows=68805 loops=3)
                    ->  Index Scan using repo_pkey on repo  (cost=0.43..1.13 rows=1 width=39) (actual time=0.004..0.004 rows=1 loops=206416)
                          Index Cond: (id = esr.repo_id)
                          Filter: (deleted_at IS NULL)
                          Rows Removed by Filter: 0
Planning Time: 0.333 ms
Execution Time: 487.791 ms
```

#### After (attempt 2)

```
Gather  (cost=59179.51..199339.22 rows=206358 width=39) (actual time=1.589..437.872 rows=206400 loops=1)
  Workers Planned: 2
  Workers Launched: 2
  ->  Nested Loop  (cost=58179.51..177703.42 rows=85982 width=39) (actual time=1.025..370.151 rows=68800 loops=3)
        ->  Hash Left Join  (cost=58179.08..63431.76 rows=85997 width=8) (actual time=0.968..67.309 rows=68805 loops=3)
              Hash Cond: (dr.repo_id = esr.repo_id)
              ->  Parallel Index Only Scan using default_repos_pkey on default_repos dr  (cost=0.42..3522.35 rows=85997 width=4) (actual time=0.040..37.065 rows=68805 loops=3)
                    Heap Fetches: 206416
              ->  Hash  (cost=55265.88..55265.88 rows=177503 width=4) (actual time=0.058..0.058 rows=0 loops=3)
                    Buckets: 131072  Batches: 4  Memory Usage: 1024kB
                    ->  Nested Loop  (cost=0.43..55265.88 rows=177503 width=4) (actual time=0.057..0.057 rows=0 loops=3)
                          ->  Seq Scan on external_services es  (cost=0.00..2.21 rows=1 width=8) (actual time=0.057..0.057 rows=0 loops=3)
                                Filter: ((namespace_user_id IS NOT NULL) AND (deleted_at IS NULL))
                                Rows Removed by Filter: 21
                          ->  Index Scan using external_service_repos_external_service_id on external_service_repos esr  (cost=0.43..45944.79 rows=931888 width=12) (never executed)
                                Index Cond: (external_service_id = es.id)
        ->  Index Scan using repo_pkey on repo  (cost=0.43..1.33 rows=1 width=39) (actual time=0.004..0.004 rows=1 loops=206416)
              Index Cond: (id = COALESCE(esr.repo_id, dr.repo_id))
              Filter: (deleted_at IS NULL)
              Rows Removed by Filter: 0
Planning Time: 2.565 ms
Execution Time: 447.685 ms
```

### CODE REVIEW

**Comment by tsenart** [comment]
@asdine, @ryanslade: We found that there are repo rows that don't have a matching row in `external_service_repos`. I don't think that is supposed to happen, right?

```
SELECT COUNT(*) FROM repo
LEFT JOIN external_service_repos esr ON repo.id = esr.repo_id
WHERE esr.repo_id IS NULL AND repo.deleted_at IS NULL;

=> 31179
```

File: internal/db/default_repos.go [hunk 1]
```diff
@@ -39,29 +39,21 @@ func (s *defaultRepos) List(ctx context.Context) (results []*types.Repo, err err
...
 	const q = `
 -- source: internal/db/default_repos.go:defaultRepos.List
-SELECT
-    id,
-    name
-FROM
-    repo r
+SELECT DISTINCT ON(repo.id) repo.id, repo.name
+FROM repo
+JOIN LATERAL (
```
Comment (unknwon) [inline internal/db/default_repos.go:44]: TIL "LATERAL". If I understand correctly, the use of "LATERAL" here is to make the subquery form a table that could be used for "JOIN"?

**Review by unknwon — APPROVED** [review]
"Execution Time" LGTM :)

**Comment by ryanslade** [comment]
> @asdine, @ryanslade: We found that there are repo rows that don't have a matching row in `external_service_repos`. I don't think that is supposed to happen, right?
> 
> ```
> SELECT COUNT(*) FROM repo
> LEFT JOIN external_service_repos esr ON repo.id = esr.repo_id
> WHERE esr.repo_id IS NULL AND repo.deleted_at IS NULL;
> 
> => 31179
> ```

Correct, that shouldn't be possible. I'm going to create an issue so that we can look into this.

**Review by ryanslade — APPROVED** [review]
Nice 👍

File: internal/db/default_repos.go (hunk 1 above)
Comment (felixge) [reply internal/db/default_repos.go:44]: I think the `LATERAL` is not needed for this query and can be removed. It's only needed when referencing previous FROM or JOIN items, which isn't the case here.

Comment (asdine) [reply internal/db/default_repos.go:44]: TIL as well!

**Review by asdine — APPROVED** [review]
Good job! 👏🏼

**Comment by tsenart** [comment]
I re-ran the new query this morning, and lo and behold, it took 3.4s! @felixge and I looked into it and found that the Postgres query planner estimate on the number of rows produced by the first sub-query was wildly wrong — so we ran `VACUUM ANALYZE external_services`, which forces an update of the query planner statistics. Once that was done, we re-ran the query and we were back to yesterday's query plan and execution time.

However, @felixge thought that this query was brittle in the face of outdated statistics, which can happen with very small tables (like external_services, for now, before we get more user added ones), since the auto vacuum feature of Postgres often deprioritises updating those in favour of larger tables.

Based on that, we chose to use an alternative query we had played around with yesterday, which while it does slightly more I/O, it's slightly faster and isn't affected by the those statistics in the same way — Postgres produced the right query plan even before us running `VACUUM ANALYZE external_services`.

My last commit pushes that query, and here's the query plan.

```
Gather  (cost=59179.51..199339.22 rows=206358 width=39) (actual time=1.589..437.872 rows=206400 loops=1)
  Workers Planned: 2
  Workers Launched: 2
  ->  Nested Loop  (cost=58179.51..177703.42 rows=85982 width=39) (actual time=1.025..370.151 rows=68800 loops=3)
        ->  Hash Left Join  (cost=58179.08..63431.76 rows=85997 width=8) (actual time=0.968..67.309 rows=68805 loops=3)
              Hash Cond: (dr.repo_id = esr.repo_id)
              ->  Parallel Index Only Scan using default_repos_pkey on default_repos dr  (cost=0.42..3522.35 rows=85997 width=4) (actual time=0.040..37.065 rows=68805 loops=3)
                    Heap Fetches: 206416
              ->  Hash  (cost=55265.88..55265.88 rows=177503 width=4) (actual time=0.058..0.058 rows=0 loops=3)
                    Buckets: 131072  Batches: 4  Memory Usage: 1024kB
                    ->  Nested Loop  (cost=0.43..55265.88 rows=177503 width=4) (actual time=0.057..0.057 rows=0 loops=3)
                          ->  Seq Scan on external_services es  (cost=0.00..2.21 rows=1 width=8) (actual time=0.057..0.057 rows=0 loops=3)
                                Filter: ((namespace_user_id IS NOT NULL) AND (deleted_at IS NULL))
                                Rows Removed by Filter: 21
                          ->  Index Scan using external_service_repos_external_service_id on external_service_repos esr  (cost=0.43..45944.79 rows=931888 width=12) (never executed)
                                Index Cond: (external_service_id = es.id)
        ->  Index Scan using repo_pkey on repo  (cost=0.43..1.33 rows=1 width=39) (actual time=0.004..0.004 rows=1 loops=206416)
              Index Cond: (id = COALESCE(esr.repo_id, dr.repo_id))
              Filter: (deleted_at IS NULL)
              Rows Removed by Filter: 0
Planning Time: 2.565 ms
Execution Time: 447.685 ms
```

**Comment by tsenart** [comment]
Ahhh, there is a bug with the last query unfortunately. Looking into it.

**Comment by tsenart** [comment]
So we're back to the query that is in master :D Unfortunately the fact that one external service has most repos really throws off the estimates of the query planner, which then produces a bad query plan. The original query seems to do relatively fine compared to the one we developed next. I just pushed a commit to add the filter for non deleted repos to the default_repos part. Merging after this!

**Comment by felixge** [comment]
@tsenart beware that the original plan's misestimate is even worse than the optimized one we were working on:

Unique  (cost=1325604.75..1351922.72 rows=**3509063** width=36) (actual time=1135.181..1410.477 rows=**206416** loops=1)

So the fact that it "relatively fine" is more of a lucky coincidence for now.

But yeah, as you said, it's difficult for PostgreSQL to estimate this query. It's smart enough to understand the irregular distribution of `external_service_repos.external_service_id`, but it can't apply this knowledge without partially evaluating the query against `external_services`. I don't know any DB that does partial query evaluation for planning purposes. IMO the best way to solve this would be to split this into two queries. The first fetches the list of `external_service_id`, and the second uses it as a static input allowing the query planner to properly estimate things. This could be done with 2 queries from Go, or a plpgsql procedure.

**Pull request merged**